### DISCUSSION TITLE
openquestions/discretization-hypothesis/

### DISCUSSION THREAD

**Discussion by giscus** [learningmechanics/learningmechanics.github.io#9] [Comments]
# openquestions/discretization-hypothesis/

Discussion page for open question: Are finite neural networks properly understood as approximations to infinite limits?

https://learningmechanics.pub/openquestions/discretization-hypothesis/

**Comment by EvergreenTree** [comment]
This is where I’d place my work too: finite-depth Transformers as discretizations of an effective-field / variational dynamics over token-feature fields.

The architecture-side question is: can attention, MLP, softmax, etc. be derived as the minimal interaction terms of the continuum limit?

https://researchseminars.org/talk/TropicalmathandML/31/

**Reply by james-simon** [reply]
Yeah, interesting to ask: supposing we identify a desired continuum process, what's the most compute-efficient way to discretize it? For example, pointwise nonlins are so widely used not bc they're mathematically nice (I claim), but bc they're computationally cheap. I suppose an affirmative answer to this Open Dir should be able to retrodict this and other standard architectural choices.

**Reply by EvergreenTree** [reply]
Great question! Pointwise activations are great both computationally, because they are easy to parallelize/fuse in kernels, and mathematically, because they preserve projection/permutation-equivariant structure. One elegance of ReLU is that it can be viewed as shifting GD toward projected GD, where the cone direction is learned by the MLP parameters.

Computationally, I designed a non-pointwise activation function [CoLU](https://proceedings.mlr.press/v285/fu24a.html), where the extra operation is only a groupwise reduction. It remains linear in width, and empirically I do not observe a train/eval throughput drop. So I think “pointwise” is a very useful discretization/implementation prior, but not necessarily a hard computational boundary.
